In [14]:
import pandas as pd

In [16]:
!ls -lah /content

total 188K
drwxr-xr-x 1 root root 4.0K Sep 12 17:29 .
drwxr-xr-x 1 root root 4.0K Sep 12 17:26 ..
drwxr-xr-x 4 root root 4.0K Sep  4 13:25 .config
-rw-r--r-- 1 root root 172K Sep 12 17:29 dataset.zip
drwxr-xr-x 1 root root 4.0K Sep  4 13:25 sample_data


In [17]:
!unzip -o "/content/dataset.zip"

Archive:  /content/dataset.zip
  inflating: Telco_Cusomer_Churn.csv  


In [18]:
!ls -lah /content

total 1.2M
drwxr-xr-x 1 root root 4.0K Sep 12 17:33 .
drwxr-xr-x 1 root root 4.0K Sep 12 17:26 ..
drwxr-xr-x 4 root root 4.0K Sep  4 13:25 .config
-rw-r--r-- 1 root root 172K Sep 12 17:29 dataset.zip
drwxr-xr-x 1 root root 4.0K Sep  4 13:25 sample_data
-rw-r--r-- 1 root root 955K Aug 21  2025 Telco_Cusomer_Churn.csv


In [19]:
df = pd.read_csv("/content/Telco_Cusomer_Churn.csv")

In [20]:
df.shape

(7043, 21)

In [21]:
df = df.drop("customerID", axis=1)

In [22]:
df.shape

(7043, 20)

In [23]:
X = df.drop("Churn", axis=1)
y = df["Churn"]

In [24]:
print(X.shape)
print(y.shape)

(7043, 19)
(7043,)


In [25]:
y = y.map({"No": 0, "Yes": 1})

In [26]:
y.value_counts()

,count
Churn,
0,5174
1,1869


In [27]:
from sklearn.model_selection import train_test_split

In [28]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

In [29]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(5634, 19)
(1409, 19)
(5634,)
(1409,)


In [30]:
X.select_dtypes(include="object").columns

Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'TotalCharges'],
      dtype='object')

In [31]:
df["TotalCharges"] = df["TotalCharges"].str.strip()

In [32]:
df["TotalCharges"] = df["TotalCharges"].replace("", "0")

In [33]:
df["TotalCharges"] = pd.to_numeric(df["TotalCharges"])

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   MultipleLines     7043 non-null   object 
 7   InternetService   7043 non-null   object 
 8   OnlineSecurity    7043 non-null   object 
 9   OnlineBackup      7043 non-null   object 
 10  DeviceProtection  7043 non-null   object 
 11  TechSupport       7043 non-null   object 
 12  StreamingTV       7043 non-null   object 
 13  StreamingMovies   7043 non-null   object 
 14  Contract          7043 non-null   object 
 15  PaperlessBilling  7043 non-null   object 
 16  PaymentMethod     7043 non-null   object 


In [36]:
X = df.drop("Churn", axis=1)

In [37]:
y = df["Churn"].map({"No": 0, "Yes": 1})

In [38]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [44]:
df.to_csv("/content/cleaned_customer_churn.csv", index=False)

In [45]:
!ls -lh /content/cleaned_customer_churn.csv

-rw-r--r-- 1 root root 874K Sep 12 18:07 /content/cleaned_customer_churn.csv


## Problem Statement

Customer churn is an important business problem because losing customers can reduce company revenue and increase the cost of acquiring new customers.

The goal of this project is to use customer information such as tenure, contract type, monthly charges, internet service, payment method, and other service-related features to predict whether a customer is likely to churn.

This is a supervised machine learning classification problem because the dataset contains a known target variable, Churn, with two possible outcomes: Yes (customer left) and No (customer stayed).

## Data Quality Report

- The dataset contains 7,043 customer records and 21 original columns.
- No duplicate records were found.
- No missing values were detected using the standard missing-value check.
- 11 blank values were found in the TotalCharges column.
- These blank TotalCharges values occurred for customers with zero months of tenure, so they were replaced with 0.
- TotalCharges was converted from text to a numeric data type.
- The customerID column was removed because it is an identifier and does not provide useful information for predicting churn.
- Categorical variables were converted into numerical features using one-hot encoding.
- The Churn target was converted from Yes/No to 1/0.
- The data was divided into 80% training data and 20% testing data.

## Target Distribution Analysis

The target variable is Churn.

- Customers who did not churn (No): 5,174
- Customers who churned (Yes): 1,869
- Total customers: 7,043

Approximately 73.5% of customers did not churn, while approximately 26.5% churned.

This shows that the target variable is imbalanced because there are more customers who stayed than customers who left.

In [47]:
# Create X and y again
X = df.drop("Churn", axis=1)
y = df["Churn"].map({"No": 0, "Yes": 1})

# Identify categorical columns
categorical_columns = X.select_dtypes(include="object").columns

print("Categorical columns:")
print(categorical_columns)

Categorical columns:
Index(['gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod'],
      dtype='object')


In [48]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

preprocessor = ColumnTransformer(
    transformers=[
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_columns)
    ],
    remainder="passthrough"
)

print("Encoder created successfully!")

Encoder created successfully!


In [49]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)

Training data: (5634, 19)
Testing data: (1409, 19)


In [50]:
X_train_encoded = preprocessor.fit_transform(X_train)
X_test_encoded = preprocessor.transform(X_test)

print("Encoded training data:", X_train_encoded.shape)
print("Encoded testing data:", X_test_encoded.shape)

Encoded training data: (5634, 45)
Encoded testing data: (1409, 45)


## Conclusion

The customer churn dataset was successfully inspected, cleaned, and prepared for machine learning.

The data was checked for missing values, duplicate records, and inconsistent values. Blank values in the TotalCharges column were replaced with 0 because they belonged to customers with zero tenure. The TotalCharges column was converted to a numeric data type.

The customerID column was removed because it is only an identifier and does not provide useful information for predicting churn.

The Churn target was converted from Yes/No to 1/0, and categorical features were converted into numerical features using one-hot encoding.

Finally, the dataset was divided into 80% training data and 20% testing data. The prepared dataset is now ready for machine learning model development.